crates/shared/src/price_estimation/competition.rs

```diff
@@ -125,12 +128,18 @@ impl HistoricalWinners {
     }
 }
 
+type PriceEstimationStage = Vec<(String, Arc<dyn PriceEstimating>)>;
+type SingleEstimatorResult = (EstimatorIndex, PriceEstimateResult);
+
 /// Price estimator that pulls estimates from various sources
-/// and competes on the best price. Returns a price estimation
-/// early if there is a configurable number of successful estimates
-/// for every query or if all price sources returned an estimate.
+/// and competes on the best price. Sources are provided as a list of lists, the
+/// outer list representing the sequential stage of the search, and the inner
+/// list representing all source that should be queried in parallel in the given
+/// stage Returns a price estimation early if there is a configurable number of
+/// successful estimates for every query or if all price sources returned an
+/// estimate.
 pub struct RacingCompetitionPriceEstimator {
-    inner: Vec<(String, Arc<dyn PriceEstimating>)>,
+    inner: Vec<PriceEstimationStage>,
```

Comment:
I think this can be achieved cleaner by introducing a `PriorityPriceEstimator` which simply checks asks every `dyn PriceEstimating` in order until it gets a price.
That way the `RacingCompetitionPriceEstimator` does not have to think about 2 separate concepts and we can easily compose the estimators like this:
```
- Priority
    - Competition1
        - cheap price feed 1
        - cheap price feed 2
    - Competition2
        - expensive price feed 1
```

Reply:
Wouldn't this require passing down information about how many estimators have succeeded in the previous stage (or reinstantiating the inner estimator with a lower threshold)?
Otherwise if stage 1 produces one result and stage 2 produces another result we may still go to stage 3 even if two results would have been enough.

Reply:
Hmm, that's true. I didn't consider the racing part. 🤔